Enable monitoring in sysgo

op-will opened the pull request:
<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**
Adds L2 observability to sysgo, wiring it up for `kona-node` and `op-reth`.

Note: observability still needs to be configured for other L2 components that expose metrics, but this PR provides a pattern to do so.

<!--
A clear and concise description of the features you're adding in this pull request.
-->

**Tests**
Added tests for utils that were added in this PR as well as a test that confirms that the option returned from `WithL2MetricsDashboard` works in the default case (does not fail when being applied).

<!--
Please describe any tests you've added. If you've added no tests, or left important behavior untested, please explain why not.
-->

**Additional context**
The new `WithL2MetricsDashboard` option factory function configures and starts dockerized prometheus and grafana instances to expose metrics if any component has registered metrics endpoints with the `Orchestrator`.

Note: This functionality is disabled by default. If enabled, docker is required to be installed, and the `SYSGO_DOCKER_EXEC_PATH` should be set to the docker binary if it is not already in your `PATH`.

<!--
Add any other context about the problem you're solving.
-->

**Metadata**
- Closes #17941 
<!-- 
Include a link to any github issues that this may close in the following form:
- Fixes #[Link to Issue]
-->

op-will committed:
WIP

- Passing metrics env vars through in l2_cl_kona
- Added L2MetricsDashboard sysgo lifecycle component

op-will committed:
Documenting and wiring up L2MetricsDashboard

Also adds some util unit tests

theochap commented on op-devstack/README.md:
```diff
@@ -163,5 +164,10 @@ Go stack env vars:
 - `SYSGO_GETH_EXEC_PATH=/path/to/geth` to select the geth executable to run
 
+L2 Observability env vars (note: only applicable if any L2 component calls `Orchestrator.RegisterL2MetricsEndpoints`):
+- `SYSGO_DOCKER_EXEC_PATH` to set local path to docker executable (defaults to `/usr/local/bin/docker`)
```
Nit: can we just assume that docker is in the path ENV var and remove that env variable? In general that's the case

op-will replied:
How do you feel about
```suggestion
- `SYSGO_DOCKER_EXEC_PATH` path to docker executable (defaults to `docker` assuming it is in `PATH`)
```

op-will replied:
Updated here: https://github.com/ethereum-optimism/optimism/pull/18017/commits/c61c6126d8c854433c85ee44e74f407d0261ef0e

theochap replied:
That works too!

op-will committed:
Adding op-reth metrics support

Also:
- Updates README with more useful metrics info
- Fixes error log in l2_cl_kona.go
- Changes default docker exec path to `docker`
- Makes available port util threadsafe

theochap commented on op-devstack/sysgo/l2_el_opreth.go:
```diff
@@ -207,53 +261,60 @@ func WithOpReth(id stack.L2ELNodeID, opts ...L2ELOption) stack.Option[*Orchestra
+			"--builder.interval=100ms",
 			"--chain=" + chainConfigPath,
-			"--with-unused-ports",
+			"--color=never",
```
good call

theochap commented on op-devstack/sysgo/system.go:
```diff
@@ -82,6 +82,8 @@ func defaultMinimalSystemOpts(ids *DefaultMinimalSystemIDs, dest *DefaultMinimal
 		ids.L2EL,
 	}))
 
+	opt.Add(WithL2MetricsDashboard())
```
Not sure the go team will want to have L2 metric dashboards by default in their presets.

op-will replied:
Right now I've added it to every preset since exposing metrics is disabled by default in the various components and only optionally enabled via env vars ([1](https://github.com/ethereum-optimism/optimism/blob/def104a0e887a98295001d20e476d0c1ab8cf8a4/op-devstack/sysgo/l2_cl_kona.go#L228) | [2](https://github.com/ethereum-optimism/optimism/blob/def104a0e887a98295001d20e476d0c1ab8cf8a4/op-devstack/sysgo/l2_el_opreth.go#L296)).

That said:
1. There's no guarantee that future components implement it this way, leading me to agree with you
2. Maybe I should have it default to enabled in the various components?

Please let me know your thoughts on 2

theochap replied:
So I don't think we would want to enable dashboards by default. Those presets are mainly used inside the go testing suite for e2e tests only so it doesn't help as much as in `op-up` (to run long-running networks) for instance.

Let's keep it that way for now and if the go team disagrees, we could remove it

sebastianst replied:
I agree, metrics dashboards should be fully disabled by default because it's mostly irrelevant to acceptance tests, where the devstack is used mostly.

op-will replied:
The dashboard component is included in this and other presets, but metrics are disabled by default (https://github.com/ethereum-optimism/optimism/commit/41d15d7b78b02eb6a5b4a00f37e4b2d6111bc9ed).

Please let me know if you'd like me to remove it from this or other presets.

joshklop replied:
I agree with the approach taken. I think keeping it in the presets is nice even if it's a no-op by default.

scharissis replied:
The approach sounds good, assuming it's truly practically a no-op and thus doesn't affect the performance of the presets and tests which will not make use of dashboards.

theochap commented on op-devstack/sysgo/util.go:
```diff
@@ -0,0 +1,52 @@
+	}
+
+	// NB: this is a hack, but it should never happen, so don't return & check for error on every call.
+	return "[should never happen] no available port found... this will err when you try to use it"
```
We should probably return an error in that case. I agree that this is an edge case (to have all the ports already taken), but if we try to allocate ports in a loop this can happen - an error message would be more helpful here.

op-will replied:
https://github.com/ethereum-optimism/optimism/pull/18017/commits/def104a0e887a98295001d20e476d0c1ab8cf8a4

theochap commented on op-up/main.go:
```diff
@@ -131,6 +131,7 @@ func runOpUp(ctx context.Context, stderr io.Writer, opUpDir string) error {
 
 		sysgo.WithL2ELNode(ids.L2EL),
 		sysgo.WithL2CLNode(ids.L2CL, ids.L1CL, ids.L1EL, ids.L2EL, sysgo.L2CLSequencer()),
+		sysgo.WithL2MetricsDashboard(),
```
Here is probably a good idea to keep this metrics dashboard (`op-up` is for long-running devnets)

sebastianst replied:
Yes, although we should probably add a flag to `op-up` to allow disabling metrics. Or is this already possible via some env var?

op-will replied:
It is already possible*, but not in a future-proof way. I'll add a global metrics enabling env var that defaults to false.

*Each individual component that can expose metrics endpoints to be scraped exposes an env var that may be set to enable it (defaulting to disabled). If no such component is enabled, prometheus and grafana will not be started. We can more directly and globally configure metrics, defaulting to false.

op-will replied:
> Yes, although we should probably add a flag to op-up to allow disabling metrics. Or is this already possible via some env var?

Metrics are disabled by default. There is now a single env var to enable metrics which defaults to disabled. See: https://github.com/ethereum-optimism/optimism/pull/18017/commits/41d15d7b78b02eb6a5b4a00f37e4b2d6111bc9ed

scharissis replied:
So, if I understand correctly, this means that `op-up` will by default NOT have metrics enabled? (As I don't see the env var `SYSGO_METRICS_ENABLED` being set to `true` anywhere here.

I believe it to be preferable for op-up to have metrics by default.

joshklop replied:
`op-up` is in a very rudimentary state right now, primarily targeting new smart contract developers. Development was paused several months ago shortly after the MVP was written. It will likely need to be overhauled (though it's scope can remain small) to meet our evolving needs.

I suggest we discuss `op-up` improvements elsewhere since this PR is more general.

scharissis replied:
Sure; no need to couple op-up to this.

op-will committed:
Updating GetAvailableLocalPort signature

sebastianst commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,325 @@
+	// Create the sub-process.
+	// We pipe sub-process logs to the test-logger.
+	logOut := logpipe.ToLogger(g.p.Logger().New("src", "stdout"))
+	logErr := logpipe.ToLogger(g.p.Logger().New("src", "stderr"))
```
Wondering if we want to add more logging context so that it's clear this is coming from prometheus, like

```suggestion
	logOut := logpipe.ToLogger(g.p.Logger().New("component", "prometheus", "src", "stdout"))
	logErr := logpipe.ToLogger(g.p.Logger().New("component", "prometheus", "src", "stderr"))
```

op-will replied:
https://github.com/ethereum-optimism/optimism/pull/18017/commits/2b5bfea54d5871e5c80dca0c3751dcf454f55c2f

sebastianst commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,325 @@
+	// Create the sub-process.
+	// We pipe sub-process logs to the test-logger.
+	logOut := logpipe.ToLogger(g.p.Logger().New("src", "stdout"))
+	logErr := logpipe.ToLogger(g.p.Logger().New("src", "stderr"))
```
Same here, maybe add `"component", "grafana"`?

op-will replied:
https://github.com/ethereum-optimism/optimism/pull/18017/commits/2b5bfea54d5871e5c80dca0c3751dcf454f55c2f

sebastianst commented on op-devstack/sysgo/util.go:
```diff
@@ -0,0 +1,52 @@
+
+// GetEnvVarOrDefault returns the value of the provided env var or the provided default value if unset or empty.
+func GetEnvVarOrDefault(envVarName string, defaultValue string) string {
+	val := os.Getenv(envVarName)
```
nit, you could use `LookupEnv` to allow setting the empty string.

op-will replied:
https://github.com/ethereum-optimism/optimism/pull/18017/commits/7ec15b74738a6a165f3904296a6b2007bdcb733a

op-will committed:
Adding logging context in kona-node, op-reth, l2 metrics dashboard

op-will committed:
GetEnvVarOrDefault: os.GetEnv => os.LookupEnv to allow for setting ""

op-will committed:
Single env var to enable/disable metrics for all configured components

joshklop commented on op-devstack/sysgo/config.go:
```diff
@@ -0,0 +1,12 @@
+
+const SysgoMetricsEnabledEnvVar = "SYSGO_METRICS_ENABLED"
+
+func AreMetricsEnabled() bool {
```
Only used inside this package, no need to export.
```suggestion
const sysgoMetricsEnabledEnvVar = "SYSGO_METRICS_ENABLED"

func areMetricsEnabled() bool {
```
(similar for the other constants, etc.)

op-will replied:
https://github.com/ethereum-optimism/optimism/pull/18017/commits/bd62f252391c74523af1228d5f7ec83e7383d9cb

Please let me know if there's anything I missed

joshklop replied:
`GetAvailableLocalPort` I think.

joshklop commented on op-devstack/sysgo/l2_el_opreth.go:
```diff
@@ -95,22 +100,34 @@ func (n *OpReth) Start() {
 			default:
 			}
+		default:
+			if endpoint := n.tryParseMetricsFromLog(msg); endpoint != nil {
```
Do you think we simplify this a bit by moving inlining the condition? The end result might look something like
```go
if msg == "RPC WS server started" {
//...
} else if msg == "RPC auth server started" {
//...
} rethURL, found := strings.CutPrefix(msg, "<metrics prefix>"); found {
// parse the reth URL
}
``` 
Same for Kona.

op-will replied:
e492f385f11fe91a44b681e879694bfd8f7088a1

joshklop commented on op-devstack/sysgo/l2_cl_kona.go:
```diff
@@ -135,6 +155,30 @@ func (k *KonaNode) InteropRPC() (endpoint string, jwtSecret eth.Bytes32) {
+//
+// If the log message doesn't appear to be metrics-related, nil will be returned. See: `konaMetricsPrefix`.
+func (k *KonaNode) tryParseMetricsFromLog(msg string) *PrometheusMetricsEndpoint {
+	if !strings.HasPrefix(msg, konaMetricsPrefix) {
```
You might find `strings.CutPrefix` useful. We should also move the `const konaMetricsPrefix` inside this function if it isn't being used elsewhere.

op-will replied:
e492f385f11fe91a44b681e879694bfd8f7088a1

joshklop commented on op-devstack/sysgo/l2_el_opreth.go:
```diff
@@ -207,53 +259,59 @@ func WithOpReth(id stack.L2ELNodeID, opts ...L2ELOption) stack.Option[*Orchestra
+		if AreMetricsEnabled() {
+			metricsPort, err := GetAvailableLocalPort()
+			p.Require().NoError(err, "WithOpReth: getting metrics port")
+			args = append(args, "--metrics="+metricsPort)
```
Since we are getting the port number from the logs anyway, can we choose `0` and avoid test flakes completely? Same for kona.

op-will replied:
Yes, although `op-reth` will just log that the port is `0`, so it doesn't work _yet_. Meant to add [this](https://github.com/ethereum-optimism/optimism/blob/41d15d7b78b02eb6a5b4a00f37e4b2d6111bc9ed/op-devstack/sysgo/l2_cl_kona.go#L252-L254) comment here as well. Will create an issue in `op-reth` and link in a comment.

op-will replied:
c6d03593f903caf8bac954e36499ccba3270ed9f

joshklop commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,325 @@
+	g.p.Require().NoError(err, "Grafana must stop")
+
+	err = g.prometheusSubprocess.Stop()
+	g.p.Require().NoError(err, "Prometheus must stop")
```
Might be nice to run these in a waitgroup for speed. Not sure if that would break something.

op-will replied:
I thought Grafana might panic if shut down second since it uses prometheus as a datasource, but I'm reading that it handles this situation gracefully so will update 👍

op-will replied:
9c98c222b371b2275f782ab8e0e165714d8a31ae

joshklop commented on op-devstack/sysgo/orchestrator.go:
```diff
@@ -18,6 +19,13 @@ import (
 	"github.com/ethereum/go-ethereum/common/hexutil"
 )
 
+type PrometheusMetricsEndpoint struct {
```
Should this go in the `l2_metrics_dashboard.go` file? Feels a bit out of place here.

op-will replied:
f755784c81485291b3160c2ea607aafffb6c25dd

joshklop commented on op-devstack/sysgo/l2_el_opreth.go:
```diff
@@ -207,53 +259,59 @@ func WithOpReth(id stack.L2ELNodeID, opts ...L2ELOption) stack.Option[*Orchestra
+			args:                     args,
+			env:                      []string{},
+			p:                        p,
+			registerMetricsEndpoints: orch.RegisterL2MetricsEndpoints,
```
Passing in callbacks as struct fields can make the code a bit hard to follow imho. Can we instead pass in the orchestrator here? Then we can remove the `p` field as well. If we want, we can use an interface with `P()` and `RegisterL2MetricsEndpoints` so there isn't any worry about modifying the orch's internal state.

op-will replied:
Left p separate since the union type didn't seem to be broadly useful. 

Please let me know what you think: dcbdf1dcf8bb44f51b22504a870e3575b27ce35d

joshklop commented on op-devstack/sysgo/orchestrator.go:
```diff
@@ -138,6 +149,14 @@ func (o *Orchestrator) Hydrate(sys stack.ExtensibleSystem) {
 
+// RegisterL2MetricsEndpoints is called by components when they are started (or earlier) to register
+// their metrics endpoints so that a prometheus instance may be spun up to scrape metrics.
+func (o *Orchestrator) RegisterL2MetricsEndpoints(serviceName string, endpoints ...PrometheusMetricsEndpoint) {
```
Can we prevent duplicates by using `stack.IDWithChain` instead of `string` for the `serviceName`?

op-will replied:
dcbdf1dcf8bb44f51b22504a870e3575b27ce35d

joshklop commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,325 @@
+			"grafana/grafana",
+		}
+		grafanaEnv := []string{
+			PropagateEnvVarOrDefault("GF_SECURITY_ADMIN_USER", "admin"),
```
Any objections to passing through all env vars with `os.Environ()` for simplicity and forward compatibility?

op-will replied:
Sorry this change got bundled with some other changes, but: d36ff0a66db409fcb69fc3260b47aa7098e7a961

I assume your suggestion is to _also_ include `os.Environ()` which makes sense

joshklop commented on op-devstack/sysgo/orchestrator.go:
```diff
@@ -18,6 +19,13 @@ import (
+	host              string
+	port              string
+	isLocal           bool
+	isRunningInDocker bool
```
Is `isLocal` ever false? Is `isRunningInDocker` ever true?

op-will replied:
Reworked here: d36ff0a66db409fcb69fc3260b47aa7098e7a961

I still think it makes sense to specify this so the dashboard can properly interpret it (and in case some components use docker or the dashboard does not in the future), but can remove `isLocal` for now and simplify via a factory function.

joshklop commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,325 @@
+		g.p.Logger().Info(fmt.Sprintf("Waiting for prometheus to start at %s...", g.prometheusEndpoint))
+
+		time.Sleep(interval)
+	}
```
Can we do something like this to avoid the `Sleep`?
```go
for {
    // do the check here
    select {
    case <-time.After(2 * time.Second):
    case <-ctx.Done():
        require.Fail("Prometheus never started")
    }
}
```
Edit: zooming out, is it necessary to wait for prometheus to start at all?

op-will replied:
It looks like grafana can start without prometheus up, so you're right, we don't need to enforce this ordering 👍

op-will replied:
9c98c222b371b2275f782ab8e0e165714d8a31ae

joshklop commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,325 @@
+	err := os.MkdirAll(dirPath, 0777)
+	p.Require().NoError(err, fmt.Sprintf("getGrafanaProvisioningDirPath: error writing dir path %s", dirPath))
+
+	p.Logger().Info(fmt.Sprintf("Created grafana/provisioning/datasources dir at %s", dirPath))
```
nit: leveraging the logger's key-value format
```suggestion
	p.Logger().Info("Created grafana/provisioning/datasources dir", "path", dirPath)
```
same for other logs if necessary

op-will replied:
e9d7e8bfa82d4054789e1f9e521b8f26c2f2db12

joshklop commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,325 @@
+	filePath := filepath.Join(dirPath, "prometheus.yml")
+	file, err := os.Create(filePath)
+	p.Require().NoError(err, fmt.Sprintf("getGrafanaProvisioningDirPath: error creating prometheus file at %s", filePath))
+	defer file.Close()
```
```suggestion
	defer func() {
	    p.Require().NoError(file.Close())
	}()
```

op-will replied:
cfe1f8dbe37851f1d8a0e06a1378efa7a5c8d3e2

joshklop replied:
Nice, didn't realize there were others.

joshklop reviewed (commented):
Really excited about this! Thank you!

op-will committed:
Encapsulate vars and functions within sysgo package

op-will committed:
Inline metrics endpoint parsing from logs

op-will committed:
Adding issues to kona and op-reth and referencing in metrics port setting

op-will committed:
Making metrics' subprocesses start/stop in parallel

op-will committed:
Moving PrometheusMetricsEndpoint into l2_metrics_dashboard.go

op-will committed:
Using new L2MetricsRegistrar interface instead of callbacks

Also uses stack.IDWithChain instead of service name

op-will committed:
Simplifying PrometheusMetricsTarget

Still conveys how to interpret the host & port without
storing that on the struct

op-will committed:
Using logger's key-value format where possible

op-will committed:
Handling file close errors when writing metrics configs

op-will committed:
Fixing broken test

op-will commented on op-devstack/sysgo/orchestrator.go:
```diff
@@ -138,6 +141,14 @@ func (o *Orchestrator) Hydrate(sys stack.ExtensibleSystem) {
 	o.sysHook.PostHydrate(sys)
 }
 
+func (o *Orchestrator) RegisterL2MetricsTargets(id stack.IDWithChain, endpoints ...PrometheusMetricsTarget) {
+	wasSet := o.l2MetricsEndpoints.SetIfMissing(id.Key(), endpoints)
+	if !wasSet {
+		existing, found := o.l2MetricsEndpoints.Get(id.Key())
+		o.p.Require().True(wasSet, "l2MetricsEndpoints endpoints not set", "key", id.Key(), "endpoints", endpoints, "conflicted", found, "conflicts", existing)
+	}
+}
```
Note: This is the intended behavior

op-will committed:
Making metrics url parsing more resilient

joshklop commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,360 @@
+}
+
+func (g *L2MetricsDashboard) Start() {
+	var startWaitGroup sync.WaitGroup
```
nit:  I don't think we need to parallelize anything here--if memory serves, starting subprocesses doesn't block in the current implementation (correct me if I'm wrong)

op-will replied:
94d3ab2f4afc2c7945249a89f2006ad7dfd8f0a0

joshklop commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,360 @@
+	}()
+
+	select {
+	case <-g.p.Ctx().Done():
```
We might be stopping the subprocesses because the context is already canceled. We should block on `Wait` to ensure all resources are cleaned up.

op-will replied:
94d3ab2f4afc2c7945249a89f2006ad7dfd8f0a0

joshklop commented on op-devstack/sysgo/util.go:
```diff
@@ -0,0 +1,52 @@
+)
+
+// GetEnvVarOrDefault returns the value of the provided env var or the provided default value if unset.
+func GetEnvVarOrDefault(envVarName string, defaultValue string) string {
```
I think the functions in this file can be un-exported.

op-will replied:
d5c972f4f039a0069005a609d2d7a735d96469ff

joshklop commented on op-devstack/sysgo/orchestrator.go:
```diff
@@ -138,6 +141,14 @@ func (o *Orchestrator) Hydrate(sys stack.ExtensibleSystem) {
+	wasSet := o.l2MetricsEndpoints.SetIfMissing(id.Key(), endpoints)
+	if !wasSet {
+		existing, found := o.l2MetricsEndpoints.Get(id.Key())
+		o.p.Require().True(wasSet, "l2MetricsEndpoints endpoints not set", "key", id.Key(), "endpoints", endpoints, "conflicted", found, "conflicts", existing)
```
Why do we need the extra check? I assume `SetIfMissing` is useful in case the same service is registered twice.

op-will replied:
The same service being registered twice feels like an error and/or non-unique names, but we don't have to err on that here

op-will replied:
Switched to logging a warning: a4b3c73e4978dcc486ca1c5100690b6f30d591c3

joshklop commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,360 @@
+
+	if err := g.grafanaSubprocess.Start(g.grafanaExecPath, g.grafanaArgs, g.grafanaEnv); err != nil {
+		g.p.Require().NoError(err, "Grafana must start")
+	}
```
```suggestion
	err := g.grafanaSubprocess.Start(g.grafanaExecPath, g.grafanaArgs, g.grafanaEnv)
	g.p.Require().NoError(err, "Grafana must start")
```

op-will replied:
e1e4ddba7764c14e75deaf36894aa6e89b817aa5

joshklop commented on op-devstack/sysgo/l2_cl_kona.go:
```diff
@@ -101,6 +102,9 @@ func (k *KonaNode) Start() {
 			userRPCChan <- "http://" + e.FieldValue("addr").(string)
 		} else if metricsUrl, found := strings.CutPrefix(msg, "Serving metrics at: "); found {
 			// Matching messages like "Serving metrics at: http://0.0.0.0:9091"
+			if !strings.HasPrefix(metricsUrl, "http") {
```
Did you see this case come up in practice?

op-will replied:
Currently `op-reth` and `kona-node` log this differently, with `op-reth` just logging `hostname:port`. In both cases, there are issues open to update logging when the passed port is `0`, so this is more of a hedge against the format changing in those PRs / future PRs.

joshklop reviewed (commented):
Looking good! Last few comments. Do you happen to have a command I can run to try it out?

joshklop reviewed (approved):
Approving optimistically with the nits previous mentioned. It works perfectly for me as-is. Awesome stuff!

op-will committed:
Start/Stop updates

- Removing goroutines from Start
- Blocking on waitgroup in Stop instead of waitgroup or context done

op-will committed:
Removing failure when registering the same endpoint twice

op-will committed:
Removing redundant if

op-will committed:
Unexporting util functions

op-will committed:
Pinning specific versions of prometheus and grafana

Overrides are possible via env vars

op-will commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,323 @@
+			"-p", fmt.Sprintf("%s:%s", grafanaServerPort, grafanaDockerPort),
+			"-v", fmt.Sprintf("%s:/etc/grafana/provisioning:ro", grafanaProvDir),
+			"-v", fmt.Sprintf("%s:/var/lib/grafana", grafanaDataDir),
+			"grafana/grafana",
```
Defaulted to a tested working tag, overridable via env vars: 1d332958e3c3b8ccbff267e8352edddaad350855

op-will commented on op-devstack/sysgo/l2_metrics_dashboard.go:
```diff
@@ -0,0 +1,323 @@
+		prometheusArgs := []string{
+			"run",
+			"-p", fmt.Sprintf("%s:%s", prometheusServerPort, prometheusDockerPort),
+			"-v", fmt.Sprintf("%s:/etc/prometheus/prometheus.yml:ro", promConfig),
+			"prom/prometheus",
+			"--config.file=/etc/prometheus/prometheus.yml",
```
Defaulted to a tested working tag, overridable via env vars: 1d332958e3c3b8ccbff267e8352edddaad350855

op-will committed:
Updating Subprocess.Stop calls after rebase
